File: sdks/php/src/QueryBuilder.php
<?php

declare(strict_types=1);

namespace Nexus\SDK;

/**
 * Fluent API for constructing Cypher queries.
 */
class QueryBuilder
{
    /** @var string[] */
    private array $matchClauses = [];

    /** @var string[] */
    private array $whereClauses = [];

    /** @var string[] */
    private array $createClauses = [];

    /** @var string[] */
    private array $setClauses = [];

    /** @var string[] */
    private array $deleteClauses = [];

    /** @var string[] */
    private array $returnClauses = [];

    /** @var string[] */
    private array $orderByClauses = [];

    private ?int $skipValue = null;
    private ?int $limitValue = null;

    /** @var array<string, mixed> */
    private array $parameters = [];

    /**
     * Create a new QueryBuilder instance.
     */
    public static function create(): self
    {
        return new self();
    }

    /**
     * Add a MATCH clause to the query.
     */
    public function match(string $pattern): self
    {
        $this->matchClauses[] = $pattern;
        return $this;
    }

    /**
     * Add an OPTIONAL MATCH clause to the query.
     */
    public function optionalMatch(string $pattern): self
    {
        $this->matchClauses[] = "OPTIONAL MATCH {$pattern}";
        return $this;
    }

    /**
     * Add a WHERE clause to the query.
     */
    public function where(string $condition): self
    {
        $this->whereClauses[] = $condition;
        return $this;
    }

    /**
     * Add an AND condition to the WHERE clause.
     */
    public function and(string $condition): self
    {
        if (count($this->whereClauses) > 0) {
            $lastIndex = count($this->whereClauses) - 1;
            $this->whereClauses[$lastIndex] .= " AND {$condition}";
        } else {
            $this->whereClauses[] = $condition;
        }
        return $this;
    }

    /**
     * Add an OR condition to the WHERE clause.
     */
    public function or(string $condition): self
    {
        if (count($this->whereClauses) > 0) {
            $lastIndex = count($this->whereClauses) - 1;
            $this->whereClauses[$lastIndex] .= " OR {$condition}";
        } else {
            $this->whereClauses[] = $condition;
        }
        return $this;
    }

    /**
     * Add a CREATE clause to the query.
     */
    public function createPattern(string $pattern): self
    {
        $this->createClauses[] = $pattern;
        return $this;
    }

    /**
     * Add a MERGE clause to the query.
     */
    public function merge(string $pattern): self
    {
        $this->createClauses[] = "MERGE {$pattern}";
        return $this;
    }

    /**
     * Add a SET clause to the query.
     */
    public function set(string $assignment): self
    {
        $this->setClauses[] = $assignment;
        return $this;
    }

    /**
     * Add a DELETE clause to the query.
     */
    public function delete(string $items): self
    {
        $this->deleteClauses[] = $items;
        return $this;
    }

    /**
     * Add a DETACH DELETE clause to the query.
     */
    public function detachDelete(string $items): self
    {
        $this->deleteClauses[] = "DETACH DELETE {$items}";
        return $this;
    }

    /**
     * Add a RETURN clause to the query.
     */
    public function return(string ...$items): self
    {
        $this->returnClauses = array_merge($this->returnClauses, $items);
        return $this;
    }

    /**
     * Add a RETURN DISTINCT clause to the query.
     */
    public function returnDistinct(string ...$items): self
    {
        if (count($this->returnClauses) === 0) {
            $this->returnClauses[] = 'DISTINCT ' . implode(', ', $items);
        } else {
            $this->returnClauses = array_merge($this->returnClauses, $items);
        }
        return $this;
    }

    /**
     * Add an ORDER BY clause to the query.
     */
    public function orderBy(string ...$items): self
    {
        $this->orderByClauses = array_merge($this->orderByClauses, $items);
        return $this;
    }

    /**
     * Add an ORDER BY ... DESC clause to the query.
     */
    public function orderByDesc(string $item): self
    {
        $this->orderByClauses[] = "{$item} DESC";
        return $this;
    }

    /**
     * Add a SKIP clause to the query.
     */
    public function skip(int $n): self
    {
        $this->skipValue = $n;
        return $this;
    }

    /**
     * Add a LIMIT clause to the query.
     */
    public function limit(int $n): self
    {
        $this->limitValue = $n;
        return $this;
    }

    /**
     * Add a parameter to the query.
     */
    public function withParam(string $name, mixed $value): self
    {
        $this->parameters[$name] = $value;
        return $this;
    }

    /**
     * Add multiple parameters to the query.
     *
     * @param array<string, mixed> $params
     */
    public function withParams(array $params): self
    {
        foreach ($params as $key => $value) {
            $this->parameters[$key] = $value;
        }
        return $this;
    }

    /**
     * Build the final Cypher query string.
     */
    public function build(): string
    {
        $parts = [];

        // MATCH clauses
        foreach ($this->matchClauses as $match) {
            if (str_starts_with($match, 'OPTIONAL MATCH')) {
                $parts[] = $match;
            } else {
                $parts[] = "MATCH {$match}";
            }
        }

        // WHERE clauses
        if (count($this->whereClauses) > 0) {
            $parts[] = 'WHERE ' . implode(' AND ', $this->whereClauses);
        }

        // CREATE/MERGE clauses
        foreach ($this->createClauses as $create) {
            if (str_starts_with($create, 'MERGE')) {
                $parts[] = $create;
            } else {
                $parts[] = "CREATE {$create}";
            }
        }

        // SET clauses
        if (count($this->setClauses) > 0) {
            $parts[] = 'SET ' . implode(', ', $this->setClauses);
        }

        // DELETE clauses
        foreach ($this->deleteClauses as $del) {
            if (str_starts_with($del, 'DETACH DELETE')) {
                $parts[] = $del;
            } else {
                $parts[] = "DELETE {$del}";
            }
        }

        // RETURN clause
        if (count($this->returnClauses) > 0) {
            $parts[] = 'RETURN ' . implode(', ', $this->returnClauses);
        }

        // ORDER BY clause
        if (count($this->orderByClauses) > 0) {
            $parts[] = 'ORDER BY ' . implode(', ', $this->orderByClauses);
        }

        // SKIP clause
        if ($this->skipValue !== null) {
            $parts[] = "SKIP {$this->skipValue}";
        }

        // LIMIT clause
        if ($this->limitValue !== null) {
            $parts[] = "LIMIT {$this->limitValue}";
        }

        return implode(' ', $parts);
    }

    /**
     * Get the parameters for the query.
     *
     * @return array<string, mixed>
     */
    public function getParameters(): array
    {
        return $this->parameters;
    }
}

/**
 * Builder for node patterns in Cypher queries.
 */
class NodePattern
{
    private string $variable;
    /** @var string[] */
    private array $labels = [];
    /** @var array<string, mixed> */
    private array $properties = [];

    public function __construct(string $variable = '')
    {
        $this->variable = $variable;
    }

    /**
     * Create a new NodePattern instance.
     */
    public static function create(string $variable = ''): self
    {
        return new self($variable);
    }

    /**
     * Set the variable name.
     */
    public function variable(string $variable): self
    {
        $this->variable = $variable;
        return $this;
    }

    /**
     * Add a label to the node.
     */
    public function withLabel(string $label): self
    {
        $this->labels[] = $label;
        return $this;
    }

    /**
     * Add multiple labels to the node.
     */
    public function withLabels(string ...$labels): self
    {
        $this->labels = array_merge($this->labels, $labels);
        return $this;
    }

    /**
     * Add a property to the node.
     */
    public function withProperty(string $key, mixed $value): self
    {
        $this->properties[$key] = $value;
        return $this;
    }

    /**
     * Add multiple properties to the node.
     *
     * @param array<string, mixed> $properties
     */
    public function withProperties(array $properties): self
    {
        foreach ($properties as $key => $value) {
            $this->properties[$key] = $value;
        }
        return $this;
    }

    /**
     * Build the node pattern string.
     */
    public function build(): string
    {
        $result = '(' . $this->variable;

        foreach ($this->labels as $label) {
            $result .= ':' . $label;
        }

        if (count($this->properties) > 0) {
            $props = [];
            foreach ($this->properties as $key => $value) {
                $props[] = "{$key}: " . self::formatValue($value);
            }
            $result .= ' {' . implode(', ', $props) . '}';
        }

        $result .= ')';
        return $result;
    }

    /**
     * Convert to string.
     */
    public function __toString(): string
    {
        return $this->build();
    }

    private static function formatValue(mixed $value): string
    {
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_string($value)) {
            return "'" . str_replace("'", "\\'", $value) . "'";
        }
        return (string) $value;
    }
}

/**
 * Builder for relationship patterns in Cypher queries.
 */
class RelationshipPattern
{
    private string $variable;
    private string $type = '';
    private string $direction = '->'; // default outgoing
    private ?int $minHops = null;
    private ?int $maxHops = null;

    public function __construct(string $variable = '')
    {
        $this->variable = $variable;
    }

    /**
     * Create a new RelationshipPattern instance.
     */
    public static function create(string $variable = ''): self
    {
        return new self($variable);
    }

    /**
     * Set the variable name.
     */
    public function variable(string $variable): self
    {
        $this->variable = $variable;
        return $this;
    }

    /**
     * Set the relationship type.
     */
    public function withType(string $type): self
    {
        $this->type = $type;
        return $this;
    }

    /**
     * Set the direction to outgoing (->).
     */
    public function outgoing(): self
    {
        $this->direction = '->';
        return $this;
    }

    /**
     * Set the direction to incoming (<-).
     */
    public function incoming(): self
    {
        $this->direction = '<-';
        return $this;
    }

    /**
     * Set the relationship to undirected (-).
     */
    public function undirected(): self
    {
        $this->direction = '';
        return $this;
    }

    /**
     * Set variable length path hops.
     */
    public function withHops(int $min, int $max): self
    {
        $this->minHops = $min;
        $this->maxHops = $max;
        return $this;
    }

    /**
     * Set minimum hops for variable length path.
     */
    public function withMinHops(int $min): self
    {
        $this->minHops = $min;
        return $this;
    }

    /**
     * Set maximum hops for variable length path.
     */
    public function withMaxHops(int $max): self
    {
        $this->maxHops = $max;
        return $this;
    }

    /**
     * Build the relationship pattern string.
     */
    public function build(): string
    {
        $result = '';

        // Start arrow
        if ($this->direction === '<-') {
            $result .= '<-[';
        } else {
            $result .= '-[';
        }

        $result .= $this->variable;

        if ($this->type !== '') {
            $result .= ':' . $this->type;
        }

        // Variable length
        if ($this->minHops !== null || $this->maxHops !== null) {
            $result .= '*';
            if ($this->minHops !== null) {
                $result .= $this->minHops;
            }
            $result .= '..';
            if ($this->maxHops !== null) {
                $result .= $this->maxHops;
            }
        }

        $result .= ']-';

        // End arrow
        if ($this->direction === '->') {
            $result .= '>';
        }

        return $result;
    }

    /**
     * Convert to string.
     */
    public function __toString(): string
    {
        return $this->build();
    }
}

/**
 * Helper class for building path patterns.
 */
class PathBuilder
{
    /**
     * Combine patterns into a path.
     */
    public static function path(string ...$patterns): string
    {
        return implode('', $patterns);
    }
}

File: sdks/csharp/QueryBuilder.cs
using System.Text;

namespace Nexus.SDK;

/// <summary>
/// Fluent API for constructing Cypher queries.
/// </summary>
public class QueryBuilder
{
    private readonly List<string> _matchClauses = new();
    private readonly List<string> _whereClauses = new();
    private readonly List<string> _createClauses = new();
    private readonly List<string> _setClauses = new();
    private readonly List<string> _deleteClauses = new();
    private readonly List<string> _returnClauses = new();
    private readonly List<string> _orderByClauses = new();
    private int? _skipValue;
    private int? _limitValue;
    private readonly Dictionary<string, object?> _parameters = new();

    /// <summary>
    /// Creates a new QueryBuilder instance.
    /// </summary>
    public static QueryBuilder Create() => new();

    /// <summary>
    /// Adds a MATCH clause to the query.
    /// </summary>
    public QueryBuilder Match(string pattern)
    {
        _matchClauses.Add(pattern);
        return this;
    }

    /// <summary>
    /// Adds an OPTIONAL MATCH clause to the query.
    /// </summary>
    public QueryBuilder OptionalMatch(string pattern)
    {
        _matchClauses.Add($"OPTIONAL MATCH {pattern}");
        return this;
    }

    /// <summary>
    /// Adds a WHERE clause to the query.
    /// </summary>
    public QueryBuilder Where(string condition)
    {
        _whereClauses.Add(condition);
        return this;
    }

    /// <summary>
    /// Adds an AND condition to the WHERE clause.
    /// </summary>
    public QueryBuilder And(string condition)
    {
        if (_whereClauses.Count > 0)
        {
            _whereClauses[^1] += $" AND {condition}";
        }
        else
        {
            _whereClauses.Add(condition);
        }
        return this;
    }

    /// <summary>
    /// Adds an OR condition to the WHERE clause.
    /// </summary>
    public QueryBuilder Or(string condition)
    {
        if (_whereClauses.Count > 0)
        {
            _whereClauses[^1] += $" OR {condition}";
        }
        else
        {
            _whereClauses.Add(condition);
        }
        return this;
    }

    /// <summary>
    /// Adds a CREATE clause to the query.
    /// </summary>
    public QueryBuilder CreatePattern(string pattern)
    {
        _createClauses.Add(pattern);
        return this;
    }

    /// <summary>
    /// Adds a MERGE clause to the query.
    /// </summary>
    public QueryBuilder Merge(string pattern)
    {
        _createClauses.Add($"MERGE {pattern}");
        return this;
    }

    /// <summary>
    /// Adds a SET clause to the query.
    /// </summary>
    public QueryBuilder Set(string assignment)
    {
        _setClauses.Add(assignment);
        return this;
    }

    /// <summary>
    /// Adds a DELETE clause to the query.
    /// </summary>
    public QueryBuilder Delete(string items)
    {
        _deleteClauses.Add(items);
        return this;
    }

    /// <summary>
    /// Adds a DETACH DELETE clause to the query.
    /// </summary>
    public QueryBuilder DetachDelete(string items)
    {
        _deleteClauses.Add($"DETACH DELETE {items}");
        return this;
    }

    /// <summary>
    /// Adds a RETURN clause to the query.
    /// </summary>
    public QueryBuilder Return(params string[] items)
    {
        _returnClauses.AddRange(items);
        return this;
    }

    /// <summary>
    /// Adds a RETURN DISTINCT clause to the query.
    /// </summary>
    public QueryBuilder ReturnDistinct(params string[] items)
    {
        if (_returnClauses.Count == 0)
        {
            _returnClauses.Add($"DISTINCT {string.Join(", ", items)}");
        }
        else
        {
            _returnClauses.AddRange(items);
        }
        return this;
    }

    /// <summary>
    /// Adds an ORDER BY clause to the query.
    /// </summary>
    public QueryBuilder OrderBy(params string[] items)
    {
        _orderByClauses.AddRange(items);
        return this;
    }

    /// <summary>
    /// Adds an ORDER BY ... DESC clause to the query.
    /// </summary>
    public QueryBuilder OrderByDesc(string item)
    {
        _orderByClauses.Add($"{item} DESC");
        return this;
    }

    /// <summary>
    /// Adds a SKIP clause to the query.
    /// </summary>
    public QueryBuilder Skip(int n)
    {
        _skipValue = n;
        return this;
    }

    /// <summary>
    /// Adds a LIMIT clause to the query.
    /// </summary>
    public QueryBuilder Limit(int n)
    {
        _limitValue = n;
        return this;
    }

    /// <summary>
    /// Adds a parameter to the query.
    /// </summary>
    public QueryBuilder WithParam(string name, object? value)
    {
        _parameters[name] = value;
        return this;
    }

    /// <summary>
    /// Adds multiple parameters to the query.
    /// </summary>
    public QueryBuilder WithParams(Dictionary<string, object?> parameters)
    {
        foreach (var (key, value) in parameters)
        {
            _parameters[key] = value;
        }
        return this;
    }

    /// <summary>
    /// Builds the final Cypher query string.
    /// </summary>
    public string Build()
    {
        var parts = new List<string>();

        // MATCH clauses
        foreach (var match in _matchClauses)
        {
            if (match.StartsWith("OPTIONAL MATCH"))
            {
                parts.Add(match);
            }
            else
            {
                parts.Add($"MATCH {match}");
            }
        }

        // WHERE clauses
        if (_whereClauses.Count > 0)
        {
            parts.Add($"WHERE {string.Join(" AND ", _whereClauses)}");
        }

        // CREATE/MERGE clauses
        foreach (var create in _createClauses)
        {
            if (create.StartsWith("MERGE"))
            {
                parts.Add(create);
            }
            else
            {
                parts.Add($"CREATE {create}");
            }
        }

        // SET clauses
        if (_setClauses.Count > 0)
        {
            parts.Add($"SET {string.Join(", ", _setClauses)}");
        }

        // DELETE clauses
        foreach (var del in _deleteClauses)
        {
            if (del.StartsWith("DETACH DELETE"))
            {
                parts.Add(del);
            }
            else
            {
                parts.Add($"DELETE {del}");
            }
        }

        // RETURN clause
        if (_returnClauses.Count > 0)
        {
            parts.Add($"RETURN {string.Join(", ", _returnClauses)}");
        }

        // ORDER BY clause
        if (_orderByClauses.Count > 0)
        {
            parts.Add($"ORDER BY {string.Join(", ", _orderByClauses)}");
        }

        // SKIP clause
        if (_skipValue.HasValue)
        {
            parts.Add($"SKIP {_skipValue}");
        }

        // LIMIT clause
        if (_limitValue.HasValue)
        {
            parts.Add($"LIMIT {_limitValue}");
        }

        return string.Join(" ", parts);
    }

    /// <summary>
    /// Gets the parameters for the query.
    /// </summary>
    public Dictionary<string, object?> Parameters => _parameters;
}

/// <summary>
/// Builder for node patterns in Cypher queries.
/// </summary>
public class NodePatternBuilder
{
    private string _variable = "";
    private readonly List<string> _labels = new();
    private readonly Dictionary<string, object?> _properties = new();

    /// <summary>
    /// Creates a new NodePatternBuilder.
    /// </summary>
    public static NodePatternBuilder Create(string variable = "")
    {
        return new NodePatternBuilder { _variable = variable };
    }

    /// <summary>
    /// Sets the variable name.
    /// </summary>
    public NodePatternBuilder Variable(string variable)
    {
        _variable = variable;
        return this;
    }

    /// <summary>
    /// Adds a label to the node.
    /// </summary>
    public NodePatternBuilder WithLabel(string label)
    {
        _labels.Add(label);
        return this;
    }

    /// <summary>
    /// Adds multiple labels to the node.
    /// </summary>
    public NodePatternBuilder WithLabels(params string[] labels)
    {
        _labels.AddRange(labels);
        return this;
    }

    /// <summary>
    /// Adds a property to the node.
    /// </summary>
    public NodePatternBuilder WithProperty(string key, object? value)
    {
        _properties[key] = value;
        return this;
    }

    /// <summary>
    /// Adds multiple properties to the node.
    /// </summary>
    public NodePatternBuilder WithProperties(Dictionary<string, object?> properties)
    {
        foreach (var (key, value) in properties)
        {
            _properties[key] = value;
        }
        return this;
    }

    /// <summary>
    /// Builds the node pattern string.
    /// </summary>
    public string Build()
    {
        var sb = new StringBuilder();
        sb.Append('(');
        sb.Append(_variable);

        foreach (var label in _labels)
        {
            sb.Append(':');
            sb.Append(label);
        }

        if (_properties.Count > 0)
        {
            sb.Append(" {");
            var first = true;
            foreach (var (key, value) in _properties)
            {
                if (!first) sb.Append(", ");
                sb.Append(key);
                sb.Append(": ");
                sb.Append(FormatValue(value));
                first = false;
            }
            sb.Append('}');
        }

        sb.Append(')');
        return sb.ToString();
    }

    /// <summary>
    /// Implicit conversion to string.
    /// </summary>
    public static implicit operator string(NodePatternBuilder builder) => builder.Build();

    private static string FormatValue(object? value)
    {
        return value switch
        {
            null => "null",
            string s => $"'{s.Replace("'", "\\'")}'",
            bool b => b ? "true" : "false",
            _ => value.ToString() ?? "null"
        };
    }
}

/// <summary>
/// Builder for relationship patterns in Cypher queries.
/// </summary>
public class RelationshipPatternBuilder
{
    private string _variable = "";
    private string _type = "";
    private RelationshipDirection _direction = RelationshipDirection.Outgoing;
    private int? _minHops;
    private int? _maxHops;

    /// <summary>
    /// Creates a new RelationshipPatternBuilder.
    /// </summary>
    public static RelationshipPatternBuilder Create(string variable = "")
    {
        return new RelationshipPatternBuilder { _variable = variable };
    }

    /// <summary>
    /// Sets the variable name.
    /// </summary>
    public RelationshipPatternBuilder Variable(string variable)
    {
        _variable = variable;
        return this;
    }

    /// <summary>
    /// Sets the relationship type.
    /// </summary>
    public RelationshipPatternBuilder WithType(string type)
    {
        _type = type;
        return this;
    }

    /// <summary>
    /// Sets the direction to outgoing (->).
    /// </summary>
    public RelationshipPatternBuilder Outgoing()
    {
        _direction = RelationshipDirection.Outgoing;
        return this;
    }

    /// <summary>
    /// Sets the direction to incoming (<-).
    /// </summary>
    public RelationshipPatternBuilder Incoming()
    {
        _direction = RelationshipDirection.Incoming;
        return this;
    }

    /// <summary>
    /// Sets the relationship to undirected (-).
    /// </summary>
    public RelationshipPatternBuilder Undirected()
    {
        _direction = RelationshipDirection.Undirected;
        return this;
    }

    /// <summary>
    /// Sets variable length path hops.
    /// </summary>
    public RelationshipPatternBuilder WithHops(int min, int max)
    {
        _minHops = min;
        _maxHops = max;
        return this;
    }

    /// <summary>
    /// Sets minimum hops for variable length path.
    /// </summary>
    public RelationshipPatternBuilder WithMinHops(int min)
    {
        _minHops = min;
        return this;
    }

    /// <summary>
    /// Sets maximum hops for variable length path.
    /// </summary>
    public RelationshipPatternBuilder WithMaxHops(int max)
    {
        _maxHops = max;
        return this;
    }

    /// <summary>
    /// Builds the relationship pattern string.
    /// </summary>
    public string Build()
    {
        var sb = new StringBuilder();

        // Start arrow
        if (_direction == RelationshipDirection.Incoming)
        {
            sb.Append("<-[");
        }
        else
        {
            sb.Append("-[");
        }

        sb.Append(_variable);

        if (!string.IsNullOrEmpty(_type))
        {
            sb.Append(':');
            sb.Append(_type);
        }

        // Variable length
        if (_minHops.HasValue || _maxHops.HasValue)
        {
            sb.Append('*');
            if (_minHops.HasValue)
            {
                sb.Append(_minHops);
            }
            sb.Append("..");
            if (_maxHops.HasValue)
            {
                sb.Append(_maxHops);
            }
        }

        sb.Append("]-");

        // End arrow
        if (_direction == RelationshipDirection.Outgoing)
        {
            sb.Append('>');
        }

        return sb.ToString();
    }

    /// <summary>
    /// Implicit conversion to string.
    /// </summary>
    public static implicit operator string(RelationshipPatternBuilder builder) => builder.Build();
}

/// <summary>
/// Direction of a relationship in a Cypher pattern.
/// </summary>
public enum RelationshipDirection
{
    /// <summary>Outgoing relationship (->)</summary>
    Outgoing,
    /// <summary>Incoming relationship (<-)</summary>
    Incoming,
    /// <summary>Undirected relationship (-)</summary>
    Undirected
}

/// <summary>
/// Helper class for building path patterns.
/// </summary>
public static class PathBuilder
{
    /// <summary>
    /// Combines patterns into a path.
    /// </summary>
    public static string Path(params string[] patterns)
    {
        return string.Concat(patterns);
    }
}
